lib/SILGen/SILGenLValue.cpp

```diff
@@ -1344,6 +1351,86 @@ namespace {
       assert(getAccessorDecl()->isSetter());
       SILDeclRef setter = Accessor;
 
+      if (hasPropertyDelegate() && IsOnSelfParameter) {
+        // This is delegated property. Instead of emitting a setter, emit an
+        // assign_by_delegate with the allocating initializer funcction and the
+        // setter function as arguments. DefiniteInitializtion will then decide
+        // between the two functions, depending if it's an initialization or a
+        // re-assignment.
+        //
+        VarDecl *field = dyn_cast<VarDecl>(Storage);
+        VarDecl *backingVar = field->getPropertyDelegateBackingVar();
+        assert(backingVar);
+        CanType ValType = backingVar->getType()->getCanonicalType();
+        SILType varStorageType =
+          SGF.SGM.Types.getSubstitutedStorageType(backingVar, ValType);
+        auto typeData =
+          getLogicalStorageTypeData(SGF.SGM, getTypeData().AccessKind, ValType);
+
+        // Get the address of the storage property.
+        ManagedValue proj;
+        if (BaseFormalType->mayHaveSuperclass()) {
+          RefElementComponent REC(backingVar, LValueOptions(), varStorageType,
+                                  typeData);
+          proj = std::move(REC).project(SGF, loc, base);
+        } else {
+          assert(BaseFormalType->getStructOrBoundGenericStruct());
```

Comment:
Can a delegate ever be implemented by an enum?

Reply:
I don't think so

Reply:
Yes, it's allowed for a delegate to be implemented by an `enum`. The `Lazy` example from the proposal is one such example.